Add functor versions of some FV BCs

GiudGiud opened the pull request:
For SANA

GiudGiud committed:
Add a functor version of the FV Thermal resistance BC, refs #16809

GiudGiud committed:
Add functor version of the FV thermal heat flux BC, refs #16809
Clang format fix on resistance BC

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/881827/) on 827877f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19347/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19347/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6dbc2add827c0f671db43acc6056c268931d392`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/881861/) on 986de98 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19347/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19347/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c6dbc2add827c0f671db43acc6056c268931d392`

lindsayad commented on modules/heat_conduction/include/fvbcs/FunctorThermalResistanceBC.h:
```diff
@@ -0,0 +1,90 @@
+protected:
+  virtual ADReal computeQpResidual() override;
+
+  void computeParallelResistance();
+
+  void computeConductionResistance();
```
doxygen

lindsayad commented on modules/heat_conduction/src/fvbcs/FunctorThermalResistanceBC.C:
```diff
@@ -0,0 +1,177 @@
+    _parallel_resistance(0.0)
+{
+  if (_k.size() != _dx.size())
+    mooseError("Number of specified thermal conductivities must match the number "
+               "of conduction layers!");
```
could this be a `paramError`?

GiudGiud replied:
yes

lindsayad commented on modules/heat_conduction/src/fvbcs/FunctorThermalResistanceBC.C:
```diff
@@ -0,0 +1,177 @@
+{
+  Real r = _inner_radius;
+
+  for (std::size_t i = 0; i < _k.size(); ++i)
```
```suggestion
  for (const auto i : index_range(_k))
```

GiudGiud replied:
ok I ll fix it in the other non-functor too

lindsayad commented on modules/heat_conduction/src/fvbcs/FunctorThermalResistanceBC.C:
```diff
@@ -0,0 +1,177 @@
+  // resistance to find the overall heat flux. For Cartesian, dividing by the
+  // 'inner_radius' has no effect, but it is required for correct normalization
+  // for cylindrical geometries.
+  flux = (_T(&_face_info->elem()) - _T_ambient) / (_conduction_resistance + _parallel_resistance) /
+         _inner_radius;
```
Does the non-functor version of this query at `_qp`? Use of `_face_info->elem()` would be problematic if this BC was on an internal boundary where the variable is subdomain restricted and only exists on the `_face_info->neighbor()` side

lindsayad replied:
You could query the face type, and set the elem to evaluate at based on that

GiudGiud replied:
well this one very much has an outer boundary feel imo, since it s assuming radiation to infinity, an outer temperature etc

But the other BC I definitely need to change
I ll fix both just in case

lindsayad commented on modules/heat_conduction/src/fvbcs/FunctorThermalResistanceBC.C:
```diff
@@ -0,0 +1,177 @@
+
+  // for Cartesian, dividing by the 'outer_radius' has no effect, but it is
+  // required for correct normalization for cylindrical geometries
+  _parallel_resistance = 1.0 / (hr + _h(&_face_info->elem())) / _outer_radius;
```
same concern as above

GiudGiud committed:
Add NSFVFunctor heat flux BC header and docs

GiudGiud committed:
Address Alex's review:
use the elem on the right side of the boundary, in case the BC is internal to the mesh
docstring
fancy for loops range
clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/886053/) on 01e0e34 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19347/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19347/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cbd92b241ef79de31019ec748be98282a5375e69`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/886154/) on 0fc02ef wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19347/site)

This comment will be updated on new commits.

GiudGiud committed:
Add a shorthand for the single sided face argument, for convenience refs #16809

GiudGiud committed:
Use face values for the thermal flux and resistance BCs, refs #16809
Zealoous clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/886607/) on 56c9dd5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19347/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19347/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cbd92b241ef79de31019ec748be98282a5375e69`

GiudGiud committed:
Tighten solve tolerance on solve test, using off-diagonal scaling to reduce residual
Still raise tolerance on the CSV, for distributed+threaded tests, as the issue remains that
absolute diffs of 1e-15 are causing rel diffs

GiudGiud committed:
Re-gold since face evaluation of the split now, instead of cell
Change findUstar to be more resilient to potential large Newton update (use abs on rel_erro)

GiudGiud committed:
Make wall velocity solve for wall function more resilient
Make local FV tests using global AD indexing

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/888403/) on ac8a63d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>181b9f</th>
      <th colspan=4>#19347 ac8a63</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>81.94%</td>
      <td>-0.00%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72041</td>
      <td>72041</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15871</td>
      <td>15876</td>
      <td>+5</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>181b9f</th>
      <th colspan=4>#19347 ac8a63</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.62%</td>
      <td>87.65%</td>
      <td>+0.03%</td>
      <td>87.80%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2888</td>
      <td>2959</td>
      <td>+71</td>
      <td>72</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>408</td>
      <td>417</td>
      <td>+9</td>
      <td>10</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage_diff/heat_conduction.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage/heat_conduction)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>181b9f</th>
      <th colspan=4>#19347 ac8a63</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>78.85%</td>
      <td>79.04%</td>
      <td>+0.19%</td>
      <td>98.81%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7099</td>
      <td>7180</td>
      <td>+81</td>
      <td>83</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1904</td>
      <td>1904</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage_diff/navier_stokes.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19347/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19347/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%
- `heat_conduction` new line coverage rate 87.80% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
@lindsayad should be ready for review now

lindsayad commented on modules/heat_conduction/src/fvbcs/FunctorThermalResistanceBC.C:
```diff
@@ -0,0 +1,183 @@
+  computeParallelResistance();
+
+  // other iteration requirements
+  Real iteration = 0;
```
```suggestion
  unsigned int iteration = 0;
```

GiudGiud replied:
how was I not getting a warning for the comparison to _max_iterations (an unsigned int) for this

lindsayad reviewed (commented):
It would be somewhat nice to get that new `FVFluxBC` method tested in the framework but I don't see an existing object that we can simply plug it into, so NBD.

Would it be easy to make a cylindrical test? that looks like that's the only missing lines for the heat conduction BC

GiudGiud commented:
so about RZ, this fvbcs is only correct if this is the inner/outer radial boundary. If it's the top of the cylinder for example, it s kind of wrong. Adding a test that is careful about that

Not that I dont want this PR merged ^^

GiudGiud committed:
Address Alex's review:
- add test for RZ geometry (careful, can only be outer or inner cylinder)
- change type of iteration count variable

GiudGiud commented:
Review should be addressed

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/889855/) on a519096 wanted to post the following:


The following coverage requirement(s) failed:
- `ray_tracing` coverage rate 92.71% is less than the required 93.0%

loganharbour reviewed (approved):
I approve with the coverage failure and the Freya failure.

Both failures are due to an old base commit. Will pass on next.
